src/dtx/dtx_common.c

```diff
@@ -663,3 +653,205 @@ dtx_handle_resend(daos_handle_t coh, daos_unit_oid_t *oid,
 		return rc >= 0 ? -DER_INVAL : rc;
 	}
 }
+
+static void
+dtx_exec_ops_comp_cb(void **arg)
+{
+	struct dtx_exec_shard_arg	*shard_arg;
+	struct dtx_exec_arg		*exec_arg;
+	uint32_t			i;
+	uint32_t			shard_cnt;
+
+	shard_arg = arg[0];
+	exec_arg = shard_arg->exec_arg;
+	shard_cnt = exec_arg->shard_cnt;
+	D_ASSERT(shard_cnt >= 1);
+	for (i = 0; i < shard_cnt; i++) {
+		shard_arg = arg[i];
+		D_ASSERT(shard_arg->exec_arg == exec_arg);
+		if (exec_arg->exec_result == 0)
+			exec_arg->exec_result = shard_arg->exec_shard_rc;
```

Comment:
I think we need preserve all the error codes for shards, or at least make sure -EINPROGRESS doesn't overwrite other error code.

Reply:
Hmm, it will only overwrite if it is 0. what do you mean "EINPROGRESS doesn't overwrite other error code" ?

Reply:
For instance, two shards, shard 1 return -EINPROGRESS first, shard 2 return -EFAULT, only -EINPROGRESS is returned, and client will retry mistakenly.

Reply:
ah, you mean collective result. right. I will update the patch. thanks.